source/event_stream.c

```diff
@@ -1541,36 +1569,875 @@ napi_value aws_napi_event_stream_client_connection_send_protocol_message(napi_en
+     */
+    struct aws_ref_count ref_count;
 
-    return NULL;
+    /*
+     * May only be accessed from within the libuv thread.  This includes stream APIs like acquire and release.
+     */
+    struct aws_event_stream_rpc_client_continuation_token *stream;
+    bool is_closed;
+
+    /*
+     * Single count ref to the JS stream object.
+     */
+    napi_ref node_event_stream_client_stream_ref;
+
+    /*
+     * Single count ref to the node external managed by the binding.
+     */
+    napi_ref node_event_stream_client_stream_external_ref;
+
+    napi_threadsafe_function on_stream_activated;
+    napi_threadsafe_function on_stream_ended;
+    napi_threadsafe_function on_stream_message;
+};
+
+static void s_aws_event_stream_client_stream_binding_on_zero(void *context) {
+    if (context == NULL) {
+        return;
+    }
+
+    struct aws_event_stream_client_stream_binding *binding = context;
+
+    AWS_CLEAN_THREADSAFE_FUNCTION(binding, on_stream_activated);
+    AWS_CLEAN_THREADSAFE_FUNCTION(binding, on_stream_ended);
+    AWS_CLEAN_THREADSAFE_FUNCTION(binding, on_stream_message);
+
+    aws_mem_release(binding->allocator, binding);
 }
 
-napi_value aws_napi_event_stream_client_stream_close(napi_env env, napi_callback_info info) {
-    (void)info;
+static struct aws_event_stream_client_stream_binding *s_aws_event_stream_client_stream_binding_acquire(
+    struct aws_event_stream_client_stream_binding *binding) {
+    if (binding == NULL) {
+        return NULL;
+    }
+
+    aws_ref_count_acquire(&binding->ref_count);
+    return binding;
+}
 
-    napi_throw_error(env, NULL, "aws_napi_event_stream_client_stream_close - NYI");
+static struct aws_event_stream_client_stream_binding *s_aws_event_stream_client_stream_binding_release(
+    struct aws_event_stream_client_stream_binding *binding) {
+    if (binding != NULL) {
+        aws_ref_count_release(&binding->ref_count);
+    }
 
     return NULL;
 }
 
-napi_value aws_napi_event_stream_client_stream_activate(napi_env env, napi_callback_info info) {
-    (void)info;
+static void s_close_stream_binding(napi_env env, struct aws_event_stream_client_stream_binding *binding) {
+    AWS_FATAL_ASSERT(env != NULL);
 
-    napi_throw_error(env, NULL, "aws_napi_event_stream_client_stream_activate - NYI");
+    binding->is_closed = true;
 
-    return NULL;
+    napi_ref node_event_stream_client_stream_external_ref = binding->node_event_stream_client_stream_external_ref;
+    binding->node_event_stream_client_stream_external_ref = NULL;
+
+    napi_ref node_event_stream_client_stream_ref = binding->node_event_stream_client_stream_ref;
+    binding->node_event_stream_client_stream_ref = NULL;
+
+    if (node_event_stream_client_stream_external_ref != NULL) {
+        napi_delete_reference(env, node_event_stream_client_stream_external_ref);
+    }
+
+    if (node_event_stream_client_stream_ref != NULL) {
+        napi_delete_reference(env, node_event_stream_client_stream_ref);
+    }
```

Comment:
Question: Could this be simplified to the following?

~~~
if (binding->node_event_stream_client_stream_ref != NULL) {
    napi_delete_reference(env, binding->node_event_stream_client_stream_ref);
    binding->node_event_stream_client_stream_ref = NULL
}
if (binding->node_event_stream_client_stream_ref != NULL) {
    napi_delete_reference(env, binding->node_event_stream_client_stream_ref);
    binding->node_event_stream_client_stream_ref = NULL
}
~~~

Not that it is bad by any means how it is currently, I'm just curious if there is a reason for using a couple local variables (maybe threading issues?). I am guessing there is probably something I'm missing and so I wanted to ask.

Reply:
Assignment to null is intentionally done first.  It may not be necessary here, but in general, you'll see similar things in aws-c-* when we erase something before possibly triggering external callbacks that could reach back in.

Reply:
Ah okay, that makes sense and is good to know! I figured there had to be a reason but was unsure.